betty/job/__init__.py

```diff
@@ -0,0 +1,129 @@
+"""
+Provide utilities for running jobs concurrently.
+"""
+
+from __future__ import annotations
+
+from abc import ABC
+from datetime import datetime
+from enum import IntEnum
+from typing import TYPE_CHECKING, Any, Generic, TypeAlias, TypeVar, final
+from uuid import uuid4
+
+from betty.cache.no_op import NoOpCache
+from betty.progress.no_op import NoOpProgress
+
+if TYPE_CHECKING:
+    from collections.abc import Awaitable, Callable
+
+    from betty.cache import Cache
+    from betty.job.scheduler import Scheduler
+    from betty.progress import Progress
+
+
+class Context:
+    """
+    A job context.
+    """
+
+    def __init__(
+        self, *, cache: Cache[Any] | None = None, progress: Progress | None = None
+    ):
+        self._id = str(uuid4())
+        self._cache = cache or NoOpCache()
+        self._start = datetime.now()
+        self._progress = progress or NoOpProgress()
+
+    @property
+    def id(self) -> str:
+        """
+        The unique job context ID.
+        """
+        return self._id
+
+    @property
+    def cache(self) -> Cache[Any]:
+        """
+        Provide a cache for this job context.
+
+        The cache is volatile and will be discarded once the job context is completed.
+        """
+        return self._cache
+
+    @property
+    def start(self) -> datetime:
+        """
+        When the job started.
+        """
+        return self._start
+
+
+_ContextCoT = TypeVar("_ContextCoT", bound=Context, covariant=True)
+JobFunction: TypeAlias = "Callable[[Job[_ContextCoT]], Awaitable[None]]"
+
+
+@final
+class Duration(IntEnum):
```

Comment:
Do we need this still? In DefaultScheduler we are planning to batch negligible jobs but issue any others just one at a time. That means a binary setting for which we won't need an enums, and a boolean will suffice.